### DISCUSSION TITLE
Support multiple streams per radio

### DISCUSSION THREAD

**Discussion by lbschenkel** [jcorporation/webradiodb#227] [General]
Hi. First of all, great project/initiative. I am/was a user of radio-browser.info, however due to the impossibility of submitting fixes to the data there the quality is slowly deteriorating. I hope yours will end up being a worthy successor, if not for quantity, then for quality.

I believe that it would benefit if your data model did not consider a stream and a radio to be one and the same. It's quite common for the same radio to have more than one stream, with different codecs and/or bitrate. A 1:N relationship between a radio and its streams would model reality better and also be more convenient in many ways:
- For example, metadata such as title/homepage/icon/genres/etc. belongs to the radio, but the codec and bitrate (you don't currently capture this, but you could) belongs to the stream. In this way data is not duplicated and always consistent.
- Consumers of this data such as a radio player would then know all the different streams for the same radio and could automatically use the best stream but fallback to others when one stream is unreachable.
- etc.

Considering the above, I feel that using the (slightly transformed) stream URL as the de-facto radio ID / primary key as it is done now is clever but will lead to some difficulties. For example, "DR P8 Jazz" is originally imported from Moode Audio, but then a better stream is found, and you include that better stream in the database, and now currently this results in a "new" radio and two duplicated entries (in reality different streams to the same radio) when searching. It would be nice if those two were exposed as two different streams for the same radio station. Similarly, if a radio changes its stream URL, from the perspective of the consumer of the data that radio "died" and a new radio exists, instead of the same radio just changing streams.

At the very least, given the current data model, it would be accurate to use the term "radio stream" everywhere and not "radio" or "radio station" in order to prevent any confusion.

I am not claiming that implementing this is trivial (figuring out what the radio identifier should be, detecting duplicate streams of the same radio for the ingestion scripts, coming up with an improved API, etc.). I am just curious to read what are your thoughts about the above. If you believe that a better data model could be a good idea, then I could help with brainstorming/design/implementation (I am a developer myself).

Anyway, keep up the good work!

**Comment by jcorporation** [comment]
At first: thanks!

I started this project because of the bad data quality of radio-browser.info and to provide a community curated database for myMPD.
Like the moode-audio radio list, I prefer quality over quantity.

Changing the data model to provide a 1:n relationship between radio stations and streams could really improve this project, but we need a backward compatible way to not break the myMPD usage of this project. To not break this, we must keep the naming of the m3u files and provide one m3u file per stream.

First idea is to add codec and bitrate to the current metadata and use the current stream url as the default stream and naming source. We can now add additional streams with codec and bitrate as metadata to the sources directory. Naming convention could be the default stream name appended by codec and bitrate.

To create a compatible index, we add an alternativeStreams key to the json structure and create for each stream a separate m3u file in the publish folder.

```
"https___liveradio_swr_de_sw331ch_swr1bw_play_mp3.m3u": {
    "Genre": [
        "Pop",
        "Rock"
    ],
    "Name": "SWR 1",
    "Image": "https___liveradio_swr_de_sw331ch_swr1bw_play_mp3.webp",
    "Homepage": "https://www.swr.de/swr1/",
    "Country": "Germany",
    "Language": "German",
    "Description": "SWR 1 Baden-Wuerttemberg",
    "StreamUri": "https://liveradio.swr.de/sw331ch/swr1bw/play.mp3",
    "Codec": "mp3",
    "Bitrate": "128",
    "alternativeStreams": {
        "https___liveradio_swr_de_sw331ch_swr1bw_play_aac.m3u": {
             "StreamUri": "https://liveradio.swr.de/sw331ch/swr1bw/play.mp3",
             "Codec": "mp3",
             "Bitrate": "128"
        }
    }
}
```

**Reply by jcorporation** [reply]
There is now an "Add alternate stream" link in the webinterface to add an additional stream uri to a webradio.

**Reply by lbschenkel** [reply]
I didn't think you would be so fast and implement something already. :-) Let me play with what you have done and I will come back with feedback.

**Reply by jcorporation** [reply]
I had some free time because of the bad weather here in germany.
